0009-S-sdk-resiliency.md

```diff
@@ -0,0 +1,115 @@
+- Applications to communicate to sidecar when there is degradation of the sidecar's health.
+
+### Goals
+- Dapr users can talk to a remote Dapr API without having to implement resiliency logic in their app.
+- System administrators don't need to have different configurations per application based on programming language, meaning the same configuration will work with every SDK.
+
+### Current Shortfalls
+- Applications need to implement resiliency (retry and timeout) on top of existing SDK.
+
+## Related Items
+
+### Related proposals 
+
+Formalizing the proposal here from [this issue](https://github.com/dapr/dapr/issues/6609).
+
+## Expectations and alternatives
+
+* What is in scope for this proposal?
+- SDKs to support a consistent (and small) set of environment variables to configure resiliency on SDKs
+- Consistent set of retriable errors for gRPC and HTTP APIs.
+
+* What is deliberately *not* in scope?
+- Circuit Breaking
+- A highly configurable spec for resiliency policies (like the CRD in runtime)
+
+
+* What alternatives have been considered, and why do they not solve the problem?
+1. Leave every SDK as-is:
+  - Undetermined behavior when sidecar is down or too slow. For example, the Java SDK simply gets stuck forever if there is no response from the sidecar (tested with ToxiProxy).
+  - Timeout and retry needs to be implemented at the user's application.
+2. Add retry only
+  - Undetermined behavior when sidecar is down or too slow. For example, the Java SDK simply gets stuck forever if there is no response from the sidecar (tested with ToxiProxy).
+3. Let each SDK decide how to handle this.
+  - Inconsistent behavior and configuration for resiliency, requiring system admins to know specifics of each SDK.
+
+* Are there any trade-offs being made? (space for time, for example)
+1. Simplification of retry policy, having an opinionated setting for most configuration.
+2. No support for Circuit Breaking or API health check prior to calling the Dapr API.
+
+* What advantages / disadvantages does this proposal have? 
+Pros:
+- Bring consistency and simple set of configuration points that work cross SDKs
+- Document expected behavior for SDKs regarding timeout and retries
+
+Cons:
+- See trade-offs mentiond above.
+
+## Implementation Details
+
+### Design
+
+* `DAPR_API_MAX_RETRIES` defines the maximum number of retries, SDKs can determine which strategy will be implemented (linear, exponential backoff, etc). `0` is the default value and means no retry (same for negative values).
+* `DAPR_API_TIMEOUT_SECONDS` defines the maximum waiting time to connect and receive a response for an HTTP or gRPC call. Defaults to `10`. `0` (or negative) are invalid values and it will default to `10` - warning to be printed in that case. This setting is the timeout for each API invocation and not the timeout of the aggregated time for retries. This setting can be used without retries.
```

Comment:
I think the timeout needs to be defined (or definable) per-request. For example:

- Requests that save state in a state store or perform service invocation can take a long time if there's lots of data
- Requests that are long-lived streams should not have a timeout (we currently have this in gRPC but we will eventually have them in HTTP/2 handlers too)

Also connected to the point above, that making this an env vars can introduce subtle bugs due to the app running in a different enviornment (like "works on my machine").

Reply:
Also, the current default is "0" so making this "10" will be a breaking change

Reply:
> Requests that are long-lived streams should not have a timeout (we currently have this in gRPC but we will eventually have them in HTTP/2 handlers too)

@ItalyPaleAle why should streams not have a timeout?

Reply:
They’re long-lived by design. For example, the Configuration API has a stream that is used to send notifications of updated configurations. This should never time out.

Reply:
I agree, but we should still have a timeout on the initial connection dial imo

Reply:
The last changes have addressed my feedback and concerns.